### ISSUE TITLE
Rate limiter feature request. On rate limit shell action.

### ISSUE THREAD

**Issue by charlieporth1** [folbricht/routedns#206]
Rate limiter feature request. On rate limit shell action. e.g. ban IP address using script with iptables or fail2ban. For abusing IP address

**Comment by folbricht** [comment]
I quite like the idea. But we can push it even further and make it more flexible. Since `rate-limiter` already has a way to define where to route limited requests (via `limit-resolver`), all we'd need is an element that receives a query and runs a shell on it. Optionally forwards the request to an upstream resolver, perhaps based on the script's exit code.
This would let you build the above, but more than that, this element could be used in other ways too, like behind routers or blocklists. It wouldn't be just for the rate-limiter but could be placed anywhere. Thoughts?

**Comment by charlieporth1** [comment]
I like it!

**Comment by cbuijs** [comment]
Could we create a group that takes any parameters like `client-ip` or `qname` (maybe others) and add them to a `blocklist` or `allowlist` (or delete) in general?

I think the recent introduced `name` parameter could be used to point out which list to add/delete an entry to?

Now you can use any event (like ratelimit) and complement lists that can be used to block/queue/whatever.

I like the "add to list" stuff, but find "shelling out" a step to deep with security concerns in mind. But not against it.

**Comment by charlieporth1** [comment]
@cbuijs I definitely agree to the security concern because DDOS is what I'm currently going through and iptables is pretty heavy without a ddos - when loading list - You would do batch options. E.g. queue 15 (completely a swag) or so items (in this case ip address) and have a shell function write a transaction

**Comment by folbricht** [comment]
Yes, the script or executable that gets called has to be implemented carefully. We can help a bit by passing parameters in fixed positions, like $1 == IP, $2 == QType, $3 == QName. A rate-limiter could be put in front of it to protect the server.

I still think this could be a useful feature to have. I'm thinking of use-cases where a client gets banned on the FW, or someone gets notified (paged) if this path is hit. Could even implement a (very slow) blocklist that way, if the executable queries some DB or file.

**Comment by cbuijs** [comment]
In this context, this might be something to look at as a feature. RouteDNS able to upload ipset sets based on answers:

https://git.zx2c4.com/ipset-dns/about/
